src/Stratis.Bitcoin.Features.SmartContracts/SmartContractFeature.cs

```diff
@@ -42,13 +46,40 @@ public sealed class SmartContractFeature : FullNodeFeature
         private readonly ILogger logger;
         private readonly Network network;
         private readonly IStateRepositoryRoot stateRoot;
+        private readonly DataFolder dataFolder;
 
-        public SmartContractFeature(IConsensusManager consensusLoop, ILoggerFactory loggerFactory, Network network, IStateRepositoryRoot stateRoot)
+        public SmartContractFeature(IConsensusManager consensusLoop, ILoggerFactory loggerFactory, Network network, IStateRepositoryRoot stateRoot, DataFolder dataFolder)
         {
             this.consensusManager = consensusLoop;
             this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
             this.network = network;
             this.stateRoot = stateRoot;
+            this.dataFolder = dataFolder;
+        }
+
+        public void InitializeState()
+        {
+            string directoryName = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+
+            // Find the latest state checkpoint file above the current tip.
+            foreach (KeyValuePair<int, CheckpointInfo> checkpoint in this.network.Checkpoints.OrderByDescending(c => c.Key))
+            {
+                if (checkpoint.Key <= this.consensusManager.Tip.Height)
+                    break;
+
+                string checkPointFileName = Path.Combine(directoryName, $"{this.network.Name}Contracts{checkpoint.Key}.zip");
+                if (File.Exists(checkPointFileName))
+                {
+                    // Un-zip "contracts" and "contractsreceipts" to sub-folders within the data folder.
+                    if (Directory.Exists(this.dataFolder.SmartContractStatePath))
+                        Directory.Delete(this.dataFolder.SmartContractStatePath, true);
+
+                    if (Directory.Exists(this.dataFolder.SmartContractStatePath + "receipts"))
+                        Directory.Delete(this.dataFolder.SmartContractStatePath + "receipts", true);
+
+                    System.IO.Compression.ZipFile.ExtractToDirectory(checkPointFileName, this.dataFolder.RootPath);
+                }
```

Comment:
Hmm, shouldn't you `break` once the most recent zip is extracted? Obviously there would typically only be a single zip packaged with a release, but breaking seems safer.

Reply:
Yep. Definitely missing a `break`. Good spot.